<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.7.1" />
<title>autopacmen.submodules.create_gecko_model_reaction_wise API documentation</title>
<meta name="description" content="modeling_create_gecko_model.py" />
<link href='https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css' rel='stylesheet'>
<link href='https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/8.0.0/sanitize.min.css' rel='stylesheet'>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet">
<style>.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{font-weight:bold}#index h4 + ul{margin-bottom:.6em}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>autopacmen.submodules.create_gecko_model_reaction_wise</code></h1>
</header>
<section id="section-intro">
<p>modeling_create_gecko_model.py</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">#!/usr/bin/env python3
#
# Copyright 2019 PSB
#
# Licensed under the Apache License, Version 2.0 (the &#34;License&#34;);
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an &#34;AS IS&#34; BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
&#34;&#34;&#34;modeling_create_gecko_model.py
&#34;&#34;&#34;

# IMPORTS
# External modules
import cobra
import copy
import math
import statistics
from typing import List, Dict
# Internal modules
from .helper_general import json_load, standardize_folder
from .helper_create_model import add_prot_pool_reaction, get_irreversible_model, get_p_measured, \
    read_enzyme_stoichiometries_xlsx, read_protein_data_xlsx


# PUBLIC FUNCTIONS
def create_gecko_model_reaction_wise(model: cobra.Model, output_sbml_name: str,
                                     project_folder: str, project_name: str, excluded_reactions: List[str]) -&gt; cobra.Model:
    &#34;&#34;&#34;Creates a GECKO model as described in

    &lt;i&gt;
    Sánchez, B. J., Zhang, C., Nilsson, A., Lahtvee, P. J., Kerkhoven, E. J., &amp; Nielsen, J. (2017).
    Improving the phenotype predictions of a yeast genome‐scale metabolic model by incorporating enzymatic
    constraints. Molecular systems biology, 13(8).
    &lt;/i&gt;

    Arguments
    ----------
    * model: cobra.Model ~ A cobra Model representation of the metabolic network. This model will
      be changed using cobrapy functions in order to add the proteomic constraints.
    * output_sbml_name: str ~ The base name of the created SBML.
    * project_folder: str ~ The folder in which the spreadsheets and JSONs with the model&#39;s supplemental
      data can be found.
    * project_name: str ~ The sMOMENTed model creation&#39;s name, which will be added at the beginning
      of the created SBML&#39;s name.
    * excluded_reactions: List[str] ~ A string list of reaction IDs (the &#39;reverse&#39; and &#39;forward&#39;
      name additions must not be added, i.e. for &#39;ACALD_forward&#39; just &#39;ACALD&#39; has to be given) to
      which no kcat shall be added. Typically used for gas exchange reactions such as &#39;CO2tex&#39;.
    &#34;&#34;&#34;
    # Standardize project folder
    project_folder = standardize_folder(project_folder)

    # This base path is the location were the generated files wil be stored
    basepath: str = project_folder + project_name

    # READ REACTIONS&lt;-&gt;KEGG ID XLSX
    protein_id_mass_mapping: Dict[str, float] = json_load(
        basepath+&#34;_protein_id_mass_mapping.json&#34;)

    # LOAD XLSX WITH PROTEIN DATA
    # Load protein data XLSX
    protein_id_concentration_mapping, p_total, unmeasured_protein_fraction, mean_saturation = \
        read_protein_data_xlsx(basepath)
    # Read enzyme kinetics xlsx
    reaction_id_gene_rules_mapping, reaction_id_gene_rules_protein_stoichiometry_mapping = \
        read_enzyme_stoichiometries_xlsx(basepath)

    # Read reaction &lt;-&gt; kcat mapping :D
    reactions_kcat_mapping_database = json_load(
        basepath + &#34;_reactions_kcat_mapping_combined.json&#34;)
    all_kcats = [x[&#34;forward&#34;] for x in reactions_kcat_mapping_database.values()] + \
                [x[&#34;reverse&#34;] for x in reactions_kcat_mapping_database.values()]
    all_kcats = [x for x in all_kcats if not math.isnan(x)]
    default_kcat = statistics.median(all_kcats)
    print(f&#34;Default kcat is: {default_kcat}&#34;)

    # GECKO :D #
    # This ID addition will be added to all reactions which are modified by this method
    id_addition = &#34;_TG_&#34;
    # Calculate p_measured
    p_measured = get_p_measured(
        protein_id_concentration_mapping, protein_id_mass_mapping)
    # Make model irreversible
    model = get_irreversible_model(model, id_addition)
    # Add prot_pool reaction
    model, prot_pool_metabolite = add_prot_pool_reaction(model, id_addition, p_total, p_measured,
                                                         unmeasured_protein_fraction, mean_saturation)

    # Add enzyme source reaction for every unmeasured protein
    for protein_id in list(protein_id_mass_mapping.keys()):
        if protein_id in list(protein_id_concentration_mapping.keys()):  # Measured
            eu = cobra.Reaction(id=id_addition+&#34;EU_&#34;+protein_id,
                                name=f&#34;Enzyme usage reaction of measured protein {protein_id}&#34;,
                                subsystem=&#34;AutoPACMEN&#34;)
            enzyme = cobra.Metabolite(id=protein_id+&#34;_met&#34;,
                                      name=f&#34;Protein {protein_id}&#34;,
                                      compartment=&#34;AutoPACMEN&#34;)

            eu.add_metabolites({enzyme: 1.0})
            eu.lower_bound = 0
            eu.upper_bound = protein_id_concentration_mapping[protein_id]
            model.add_reactions([eu])
        else:  # Unmeasured
            er = cobra.Reaction(id=id_addition+&#34;ER_&#34;+protein_id,
                                name=f&#34;Enzyme usage reaction of unmeasured protein {protein_id}&#34;,
                                subsystem=&#34;AutoPACMEN&#34;)
            enzyme = cobra.Metabolite(id=protein_id+&#34;_met&#34;,
                                      name=f&#34;Protein {protein_id}&#34;,
                                      compartment=&#34;AutoPACMEN&#34;)
            # Mapping is in Da, GECKO uses kDa (g/mmol)
            molecular_weight = protein_id_mass_mapping[protein_id] / 1000
            er.add_metabolites({prot_pool_metabolite: -molecular_weight,
                                enzyme: 1})
            er.lower_bound = 0
            er.upper_bound = 1000.0
            model.add_reactions([er])

    # Add enzymes to reactions
    current_arm_reaction = 1
    model_reaction_ids = [x.id for x in model.reactions]
    for model_reaction_id in model_reaction_ids:
        reaction = model.reactions.get_by_id(model_reaction_id)
        splitted_id = reaction.id.split(id_addition)

        # If the reaction has no name, ignore it
        if splitted_id[0] == &#34;&#34;:
            continue
        # Take the reaction ID from the first part of the split
        reaction_id = splitted_id[0]
        # If the reaction has no associated enzyme stoichiometries, ignore it
        if reaction_id not in list(reaction_id_gene_rules_mapping.keys()):
            continue
        # If the reaction has no gene rule, ignore it
        gene_rule = reaction_id_gene_rules_mapping[reaction_id]
        if gene_rule == [&#34;&#34;]:
            continue
        # If the reaction is manually excluded, ignore it
        if reaction_id in excluded_reactions:
            continue

        all_available = True
        for enzyme in gene_rule:
            if type(enzyme) == str:
                try:
                    model.metabolites.get_by_id(enzyme+&#34;_met&#34;)
                except Exception:
                    all_available = False
                    break
            else:
                for enzyme_id in enzyme:
                    try:
                        model.metabolites.get_by_id(enzyme_id+&#34;_met&#34;)
                    except Exception:
                        all_available = False
                        break
        if not all_available:
            continue

        # Retrieve the reaction&#39;s forward and reverse kcats from the given reaction&lt;-&gt;kcat database
        if reaction_id in reactions_kcat_mapping_database.keys():
            forward_kcat = reactions_kcat_mapping_database[reaction_id][&#34;forward&#34;]
            reverse_kcat = reactions_kcat_mapping_database[reaction_id][&#34;reverse&#34;]
        # If the reaction is not in the database, set the default kcat
        else:
            forward_kcat = default_kcat
            reverse_kcat = default_kcat

        # If the given reaction&lt;-&gt;kcat database contains math.nan as the reaction&#39;s kcat,
        # set the default kcat as math.nan means that no kcat could be found.
        if math.isnan(forward_kcat):
            forward_kcat = default_kcat
        if math.isnan(reverse_kcat):
            reverse_kcat = default_kcat

        # Add the given forward or reverse kcat is the reaction was
        # splitted due to its reversibility.
        # If the reaction is not splitted, add the forward kcat (this
        # is the only possible direction for non-splitted=non-reversible
        # reactions)
        if model_reaction_id.endswith(id_addition + &#34;forward&#34;):
            reaction_kcat = forward_kcat
        elif model_reaction_id.endswith(id_addition + &#34;reverse&#34;):
            reaction_kcat = reverse_kcat
        else:
            reaction_kcat = forward_kcat

        # Add arm reaction if isozymes occur
        if len(gene_rule) &gt; 1:  # Isozymes occur :O
            arm_reaction_id = id_addition + \
                f&#34;arm_reaction_{current_arm_reaction}&#34;
            arm_reaction_name = f&#34;Arm reaction no. {current_arm_reaction} for gene rule {str(gene_rule)}&#34;
            arm_reaction = cobra.Reaction(id=arm_reaction_id,
                                          name=arm_reaction_name,
                                          subsystem=&#34;AutoPACMEN&#34;)

            arm_reaction_metabolites = {}
            for metabolite in list(reaction.metabolites.keys()):
                stoichiometry = reaction.metabolites[metabolite]
                if stoichiometry &lt; 0:  # Educt
                    arm_reaction_metabolites[metabolite] = stoichiometry
                    reaction.add_metabolites({metabolite: -stoichiometry})

            im_id = f&#34;im_{current_arm_reaction}&#34;
            im_name = f&#34;Intermediate metabolite of arm reaction {current_arm_reaction}&#34;
            intermediate_metabolite = cobra.Metabolite(id=im_id,
                                                       name=im_name,
                                                       compartment=&#34;AutoPACMEN&#34;)
            arm_reaction_metabolites[intermediate_metabolite] = 1
            arm_reaction.add_metabolites(arm_reaction_metabolites)
            reaction.add_metabolites({intermediate_metabolite: -1})

            arm_reaction.lower_bound = 0
            arm_reaction.upper_bound = reaction.upper_bound

            model.add_reactions([arm_reaction])

            current_arm_reaction += 1

        # Add reactions depending on isozyme complex presence
        new_reactions = []
        i = 1
        for isozyme_id in gene_rule:
            new_reaction = copy.deepcopy(reaction)
            new_reaction.id = new_reaction.id + id_addition + str(i)
            protein_ids = []
            if type(isozyme_id) is str:  # No complex :O
                protein = model.metabolites.get_by_id(isozyme_id+&#34;_met&#34;)
                reaction_id = reaction_id.split(&#34;_TG_&#34;)[0]

                stoichiometry = reaction_id_gene_rules_protein_stoichiometry_mapping[
                    reaction_id][isozyme_id][isozyme_id]
                stoichiometry /= (reaction_kcat * 3600)
                stoichiometry *= -1

                metabolites = {}
                metabolites[protein] = stoichiometry
                protein_ids.append(isozyme_id)
                new_reaction.add_metabolites(metabolites)
            else:  # Complex :O
                metabolites = {}
                isozyme_id = tuple(isozyme_id)

                for single_id in isozyme_id:
                    protein = model.metabolites.get_by_id(single_id+&#34;_met&#34;)
                    reaction_id = reaction_id.split(&#34;_TG_&#34;)[0]

                    stoichiometry = reaction_id_gene_rules_protein_stoichiometry_mapping[
                        reaction_id][isozyme_id][single_id]
                    stoichiometry /= (reaction_kcat * 3600)
                    stoichiometry *= -1

                    metabolites[protein] = stoichiometry
                    protein_ids.append(single_id)
                new_reaction.add_metabolites(metabolites)

            gene_reaction_rule = &#34; and &#34;.join(protein_ids)
            new_reaction.gene_reaction_rule = gene_reaction_rule
            new_reactions.append(new_reaction)
            i += 1
        model.add_reactions(new_reactions)
        model.remove_reactions([reaction])

    cobra.io.write_sbml_model(model, project_folder+output_sbml_name)

    return model


def create_gecko_model_reaction_wise_with_sbml(input_sbml_path: str, output_sbml_name: str,
                                               project_folder: str, project_name: str, excluded_reactions: List[str]) -&gt; cobra.Model:
    &#34;&#34;&#34;See create_gecko_model_reaction_wise()&#34;&#34;&#34;
    # LOAD SBML MODEL
    model: cobra.Model = cobra.io.read_sbml_model(input_sbml_path)
    # Call gecko model creation function :D
    model = create_gecko_model_reaction_wise(model, output_sbml_name,
                                             project_folder, project_name,
                                             excluded_reactions)
    return model</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="autopacmen.submodules.create_gecko_model_reaction_wise.create_gecko_model_reaction_wise"><code class="name flex">
<span>def <span class="ident">create_gecko_model_reaction_wise</span></span>(<span>model, output_sbml_name, project_folder, project_name, excluded_reactions)</span>
</code></dt>
<dd>
<section class="desc"><p>Creates a GECKO model as described in</p>
<p><i>
Sánchez, B. J., Zhang, C., Nilsson, A., Lahtvee, P. J., Kerkhoven, E. J., &amp; Nielsen, J. (2017).
Improving the phenotype predictions of a yeast genome‐scale metabolic model by incorporating enzymatic
constraints. Molecular systems biology, 13(8).
</i></p>
<h2 id="arguments">Arguments</h2>
<ul>
<li>model: cobra.Model ~ A cobra Model representation of the metabolic network. This model will
be changed using cobrapy functions in order to add the proteomic constraints.</li>
<li>output_sbml_name: str ~ The base name of the created SBML.</li>
<li>project_folder: str ~ The folder in which the spreadsheets and JSONs with the model's supplemental
data can be found.</li>
<li>project_name: str ~ The sMOMENTed model creation's name, which will be added at the beginning
of the created SBML's name.</li>
<li>excluded_reactions: List[str] ~ A string list of reaction IDs (the 'reverse' and 'forward'
name additions must not be added, i.e. for 'ACALD_forward' just 'ACALD' has to be given) to
which no kcat shall be added. Typically used for gas exchange reactions such as 'CO2tex'.</li>
</ul></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def create_gecko_model_reaction_wise(model: cobra.Model, output_sbml_name: str,
                                     project_folder: str, project_name: str, excluded_reactions: List[str]) -&gt; cobra.Model:
    &#34;&#34;&#34;Creates a GECKO model as described in

    &lt;i&gt;
    Sánchez, B. J., Zhang, C., Nilsson, A., Lahtvee, P. J., Kerkhoven, E. J., &amp; Nielsen, J. (2017).
    Improving the phenotype predictions of a yeast genome‐scale metabolic model by incorporating enzymatic
    constraints. Molecular systems biology, 13(8).
    &lt;/i&gt;

    Arguments
    ----------
    * model: cobra.Model ~ A cobra Model representation of the metabolic network. This model will
      be changed using cobrapy functions in order to add the proteomic constraints.
    * output_sbml_name: str ~ The base name of the created SBML.
    * project_folder: str ~ The folder in which the spreadsheets and JSONs with the model&#39;s supplemental
      data can be found.
    * project_name: str ~ The sMOMENTed model creation&#39;s name, which will be added at the beginning
      of the created SBML&#39;s name.
    * excluded_reactions: List[str] ~ A string list of reaction IDs (the &#39;reverse&#39; and &#39;forward&#39;
      name additions must not be added, i.e. for &#39;ACALD_forward&#39; just &#39;ACALD&#39; has to be given) to
      which no kcat shall be added. Typically used for gas exchange reactions such as &#39;CO2tex&#39;.
    &#34;&#34;&#34;
    # Standardize project folder
    project_folder = standardize_folder(project_folder)

    # This base path is the location were the generated files wil be stored
    basepath: str = project_folder + project_name

    # READ REACTIONS&lt;-&gt;KEGG ID XLSX
    protein_id_mass_mapping: Dict[str, float] = json_load(
        basepath+&#34;_protein_id_mass_mapping.json&#34;)

    # LOAD XLSX WITH PROTEIN DATA
    # Load protein data XLSX
    protein_id_concentration_mapping, p_total, unmeasured_protein_fraction, mean_saturation = \
        read_protein_data_xlsx(basepath)
    # Read enzyme kinetics xlsx
    reaction_id_gene_rules_mapping, reaction_id_gene_rules_protein_stoichiometry_mapping = \
        read_enzyme_stoichiometries_xlsx(basepath)

    # Read reaction &lt;-&gt; kcat mapping :D
    reactions_kcat_mapping_database = json_load(
        basepath + &#34;_reactions_kcat_mapping_combined.json&#34;)
    all_kcats = [x[&#34;forward&#34;] for x in reactions_kcat_mapping_database.values()] + \
                [x[&#34;reverse&#34;] for x in reactions_kcat_mapping_database.values()]
    all_kcats = [x for x in all_kcats if not math.isnan(x)]
    default_kcat = statistics.median(all_kcats)
    print(f&#34;Default kcat is: {default_kcat}&#34;)

    # GECKO :D #
    # This ID addition will be added to all reactions which are modified by this method
    id_addition = &#34;_TG_&#34;
    # Calculate p_measured
    p_measured = get_p_measured(
        protein_id_concentration_mapping, protein_id_mass_mapping)
    # Make model irreversible
    model = get_irreversible_model(model, id_addition)
    # Add prot_pool reaction
    model, prot_pool_metabolite = add_prot_pool_reaction(model, id_addition, p_total, p_measured,
                                                         unmeasured_protein_fraction, mean_saturation)

    # Add enzyme source reaction for every unmeasured protein
    for protein_id in list(protein_id_mass_mapping.keys()):
        if protein_id in list(protein_id_concentration_mapping.keys()):  # Measured
            eu = cobra.Reaction(id=id_addition+&#34;EU_&#34;+protein_id,
                                name=f&#34;Enzyme usage reaction of measured protein {protein_id}&#34;,
                                subsystem=&#34;AutoPACMEN&#34;)
            enzyme = cobra.Metabolite(id=protein_id+&#34;_met&#34;,
                                      name=f&#34;Protein {protein_id}&#34;,
                                      compartment=&#34;AutoPACMEN&#34;)

            eu.add_metabolites({enzyme: 1.0})
            eu.lower_bound = 0
            eu.upper_bound = protein_id_concentration_mapping[protein_id]
            model.add_reactions([eu])
        else:  # Unmeasured
            er = cobra.Reaction(id=id_addition+&#34;ER_&#34;+protein_id,
                                name=f&#34;Enzyme usage reaction of unmeasured protein {protein_id}&#34;,
                                subsystem=&#34;AutoPACMEN&#34;)
            enzyme = cobra.Metabolite(id=protein_id+&#34;_met&#34;,
                                      name=f&#34;Protein {protein_id}&#34;,
                                      compartment=&#34;AutoPACMEN&#34;)
            # Mapping is in Da, GECKO uses kDa (g/mmol)
            molecular_weight = protein_id_mass_mapping[protein_id] / 1000
            er.add_metabolites({prot_pool_metabolite: -molecular_weight,
                                enzyme: 1})
            er.lower_bound = 0
            er.upper_bound = 1000.0
            model.add_reactions([er])

    # Add enzymes to reactions
    current_arm_reaction = 1
    model_reaction_ids = [x.id for x in model.reactions]
    for model_reaction_id in model_reaction_ids:
        reaction = model.reactions.get_by_id(model_reaction_id)
        splitted_id = reaction.id.split(id_addition)

        # If the reaction has no name, ignore it
        if splitted_id[0] == &#34;&#34;:
            continue
        # Take the reaction ID from the first part of the split
        reaction_id = splitted_id[0]
        # If the reaction has no associated enzyme stoichiometries, ignore it
        if reaction_id not in list(reaction_id_gene_rules_mapping.keys()):
            continue
        # If the reaction has no gene rule, ignore it
        gene_rule = reaction_id_gene_rules_mapping[reaction_id]
        if gene_rule == [&#34;&#34;]:
            continue
        # If the reaction is manually excluded, ignore it
        if reaction_id in excluded_reactions:
            continue

        all_available = True
        for enzyme in gene_rule:
            if type(enzyme) == str:
                try:
                    model.metabolites.get_by_id(enzyme+&#34;_met&#34;)
                except Exception:
                    all_available = False
                    break
            else:
                for enzyme_id in enzyme:
                    try:
                        model.metabolites.get_by_id(enzyme_id+&#34;_met&#34;)
                    except Exception:
                        all_available = False
                        break
        if not all_available:
            continue

        # Retrieve the reaction&#39;s forward and reverse kcats from the given reaction&lt;-&gt;kcat database
        if reaction_id in reactions_kcat_mapping_database.keys():
            forward_kcat = reactions_kcat_mapping_database[reaction_id][&#34;forward&#34;]
            reverse_kcat = reactions_kcat_mapping_database[reaction_id][&#34;reverse&#34;]
        # If the reaction is not in the database, set the default kcat
        else:
            forward_kcat = default_kcat
            reverse_kcat = default_kcat

        # If the given reaction&lt;-&gt;kcat database contains math.nan as the reaction&#39;s kcat,
        # set the default kcat as math.nan means that no kcat could be found.
        if math.isnan(forward_kcat):
            forward_kcat = default_kcat
        if math.isnan(reverse_kcat):
            reverse_kcat = default_kcat

        # Add the given forward or reverse kcat is the reaction was
        # splitted due to its reversibility.
        # If the reaction is not splitted, add the forward kcat (this
        # is the only possible direction for non-splitted=non-reversible
        # reactions)
        if model_reaction_id.endswith(id_addition + &#34;forward&#34;):
            reaction_kcat = forward_kcat
        elif model_reaction_id.endswith(id_addition + &#34;reverse&#34;):
            reaction_kcat = reverse_kcat
        else:
            reaction_kcat = forward_kcat

        # Add arm reaction if isozymes occur
        if len(gene_rule) &gt; 1:  # Isozymes occur :O
            arm_reaction_id = id_addition + \
                f&#34;arm_reaction_{current_arm_reaction}&#34;
            arm_reaction_name = f&#34;Arm reaction no. {current_arm_reaction} for gene rule {str(gene_rule)}&#34;
            arm_reaction = cobra.Reaction(id=arm_reaction_id,
                                          name=arm_reaction_name,
                                          subsystem=&#34;AutoPACMEN&#34;)

            arm_reaction_metabolites = {}
            for metabolite in list(reaction.metabolites.keys()):
                stoichiometry = reaction.metabolites[metabolite]
                if stoichiometry &lt; 0:  # Educt
                    arm_reaction_metabolites[metabolite] = stoichiometry
                    reaction.add_metabolites({metabolite: -stoichiometry})

            im_id = f&#34;im_{current_arm_reaction}&#34;
            im_name = f&#34;Intermediate metabolite of arm reaction {current_arm_reaction}&#34;
            intermediate_metabolite = cobra.Metabolite(id=im_id,
                                                       name=im_name,
                                                       compartment=&#34;AutoPACMEN&#34;)
            arm_reaction_metabolites[intermediate_metabolite] = 1
            arm_reaction.add_metabolites(arm_reaction_metabolites)
            reaction.add_metabolites({intermediate_metabolite: -1})

            arm_reaction.lower_bound = 0
            arm_reaction.upper_bound = reaction.upper_bound

            model.add_reactions([arm_reaction])

            current_arm_reaction += 1

        # Add reactions depending on isozyme complex presence
        new_reactions = []
        i = 1
        for isozyme_id in gene_rule:
            new_reaction = copy.deepcopy(reaction)
            new_reaction.id = new_reaction.id + id_addition + str(i)
            protein_ids = []
            if type(isozyme_id) is str:  # No complex :O
                protein = model.metabolites.get_by_id(isozyme_id+&#34;_met&#34;)
                reaction_id = reaction_id.split(&#34;_TG_&#34;)[0]

                stoichiometry = reaction_id_gene_rules_protein_stoichiometry_mapping[
                    reaction_id][isozyme_id][isozyme_id]
                stoichiometry /= (reaction_kcat * 3600)
                stoichiometry *= -1

                metabolites = {}
                metabolites[protein] = stoichiometry
                protein_ids.append(isozyme_id)
                new_reaction.add_metabolites(metabolites)
            else:  # Complex :O
                metabolites = {}
                isozyme_id = tuple(isozyme_id)

                for single_id in isozyme_id:
                    protein = model.metabolites.get_by_id(single_id+&#34;_met&#34;)
                    reaction_id = reaction_id.split(&#34;_TG_&#34;)[0]

                    stoichiometry = reaction_id_gene_rules_protein_stoichiometry_mapping[
                        reaction_id][isozyme_id][single_id]
                    stoichiometry /= (reaction_kcat * 3600)
                    stoichiometry *= -1

                    metabolites[protein] = stoichiometry
                    protein_ids.append(single_id)
                new_reaction.add_metabolites(metabolites)

            gene_reaction_rule = &#34; and &#34;.join(protein_ids)
            new_reaction.gene_reaction_rule = gene_reaction_rule
            new_reactions.append(new_reaction)
            i += 1
        model.add_reactions(new_reactions)
        model.remove_reactions([reaction])

    cobra.io.write_sbml_model(model, project_folder+output_sbml_name)

    return model</code></pre>
</details>
</dd>
<dt id="autopacmen.submodules.create_gecko_model_reaction_wise.create_gecko_model_reaction_wise_with_sbml"><code class="name flex">
<span>def <span class="ident">create_gecko_model_reaction_wise_with_sbml</span></span>(<span>input_sbml_path, output_sbml_name, project_folder, project_name, excluded_reactions)</span>
</code></dt>
<dd>
<section class="desc"><p>See create_gecko_model_reaction_wise()</p></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def create_gecko_model_reaction_wise_with_sbml(input_sbml_path: str, output_sbml_name: str,
                                               project_folder: str, project_name: str, excluded_reactions: List[str]) -&gt; cobra.Model:
    &#34;&#34;&#34;See create_gecko_model_reaction_wise()&#34;&#34;&#34;
    # LOAD SBML MODEL
    model: cobra.Model = cobra.io.read_sbml_model(input_sbml_path)
    # Call gecko model creation function :D
    model = create_gecko_model_reaction_wise(model, output_sbml_name,
                                             project_folder, project_name,
                                             excluded_reactions)
    return model</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="autopacmen.submodules" href="index.html">autopacmen.submodules</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="autopacmen.submodules.create_gecko_model_reaction_wise.create_gecko_model_reaction_wise" href="#autopacmen.submodules.create_gecko_model_reaction_wise.create_gecko_model_reaction_wise">create_gecko_model_reaction_wise</a></code></li>
<li><code><a title="autopacmen.submodules.create_gecko_model_reaction_wise.create_gecko_model_reaction_wise_with_sbml" href="#autopacmen.submodules.create_gecko_model_reaction_wise.create_gecko_model_reaction_wise_with_sbml">create_gecko_model_reaction_wise_with_sbml</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.7.1</a>.</p>
</footer>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlightingOnLoad()</script>
</body>
</html>